// react-map-gl/dist/stateful-map.react.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _class, _desc, _value, _class2; // Copyright (c) 2015 Uber Technologies, Inc.

// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:

// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.

// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.


var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _autobindDecorator = require('autobind-decorator');

var _autobindDecorator2 = _interopRequireDefault(_autobindDecorator);

var _pureRenderDecorator = require('pure-render-decorator');

var _pureRenderDecorator2 = _interopRequireDefault(_pureRenderDecorator);

var _map = require('./map.react');

var _map2 = _interopRequireDefault(_map);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

function _applyDecoratedDescriptor(target, property, decorators, descriptor, context) {
  var desc = {};
  Object['ke' + 'ys'](descriptor).forEach(function (key) {
    desc[key] = descriptor[key];
  });
  desc.enumerable = !!desc.enumerable;
  desc.configurable = !!desc.configurable;

  if ('value' in desc || desc.initializer) {
    desc.writable = true;
  }

  desc = decorators.slice().reverse().reduce(function (desc, decorator) {
    return decorator(target, property, desc) || desc;
  }, desc);

  if (context && desc.initializer !== void 0) {
    desc.value = desc.initializer ? desc.initializer.call(context) : void 0;
    desc.initializer = undefined;
  }

  if (desc.initializer === void 0) {
    Object['define' + 'Property'](target, property, desc);
    desc = null;
  }

  return desc;
}

function noop() {}

var PUBLIC_MAP_STATE_DEFAULTS = {
  latitude: 0,
  longitude: 0,
  zoom: 11,
  pitch: 0,
  bearing: 0,
  altitude: 1.5
};

function compareProps(refObject, newObject) {
  for (var field in refObject) {
    if (!(field in newObject) || refObject[field] !== newObject[field]) {
      return false;
    }
  }
  return true;
}

var PROP_TYPES = {
  /**
    * `onChangeViewport` callback is fired when the user interacted with the
    * map. The object passed to the callback contains `latitude`,
    * `longitude`, `zoom` and additional information.
    */
  onChangeViewport: _react.PropTypes.func
};

var DEFAULT_PROPS = {
  onChangeViewport: noop
};

var StatefulMapGL = (0, _pureRenderDecorator2.default)(_class = (_class2 = function (_Component) {
  _inherits(StatefulMapGL, _Component);

  function StatefulMapGL(props) {
    _classCallCheck(this, StatefulMapGL);

    var _this = _possibleConstructorReturn(this, Object.getPrototypeOf(StatefulMapGL).call(this, props));

    _this.state = _this._getInternalMapState(props);
    return _this;
  }

  // Extract all public map state fields from an object


  _createClass(StatefulMapGL, [{
    key: '_getPublicMapState',
    value: function _getPublicMapState(props) {
      return _extends({}, PUBLIC_MAP_STATE_DEFAULTS, {
        latitude: props.latitude,
        longitude: props.longitude,
        zoom: props.zoom,
        pitch: props.pitch,
        bearing: props.bearing,
        altitude: props.altitude
      });
    }

    // Extract all non-public map state fields from an object

  }, {
    key: '_getInternalMapState',
    value: function _getInternalMapState(props) {
      return {
        isDragging: props.isDragging || false,
        startDragLngLat: props.startDragLngLat || null,
        startBearing: props.startBearing || null,
        startPitch: props.startPitch || null
      };
    }

    // Save internal props to state
    // Call apps onChangeViewport with public props only
    // Only call app if public props have changed.

  }, {
    key: '_onChangeViewport',
    value: function _onChangeViewport(mapState) {
      // Update state (triggering map redraw) only if internal props changed
      var internalMapState = this._getInternalMapState(mapState);
      if (!compareProps(internalMapState, this.state)) {
        this.setState(internalMapState);
      }

      // Update app (presumably triggering redraw) only if map changed vs props
      var publicMapState = this._getPublicMapState(mapState);
      if (!compareProps(publicMapState, this.props)) {
        this.props.onChangeViewport(publicMapState);
      }
    }
  }, {
    key: 'render',
    value: function render() {
      return _react2.default.createElement(_map2.default, _extends({}, this.props, this.state, {
        onChangeViewport: this._onChangeViewport }));
    }
  }]);

  return StatefulMapGL;
}(_react.Component), (_applyDecoratedDescriptor(_class2.prototype, '_onChangeViewport', [_autobindDecorator2.default], Object.getOwnPropertyDescriptor(_class2.prototype, '_onChangeViewport'), _class2.prototype)), _class2)) || _class;

exports.default = StatefulMapGL;


StatefulMapGL.propTypes = PROP_TYPES;
StatefulMapGL.defaultProps = DEFAULT_PROPS;
